Universal window rules

yshui opened the pull request:
<!-- Please enable "Allow edits from maintainers" so we can make necessary changes to your PR -->
See #1284

yshui committed:
wm/win: separate per-window options into their own struct

It's easier to just store them together, since the same group of
information is needed for parsed window rules, and for window rendering.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
win: update window opacity only when necessary

Opacity was calculated once per mapped window per frame in
`win_process_animation_and_state_change`. This is fine because
calucating it is cheap. However with upcoming changes it might become
more expensive.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
win: remove "focused" from window options

It's only used as an intermediate value for calculating other window
options, and is not used outside `win_on_factor_changed`, so there is no
reason to save it.

This also removes the "focused_force" option from the dbus interface,
which I believe nobody is using anyways.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
win: allow different dim level per window

Instead of just a yes or no. Allow for more flexibility when universal
rules are implemented.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
c2: deprecate matching against "id", "leader", and "client"

From the user's perspective they are just random, meaningless numbers.
Doesn't really make sense to write rule against them.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
wm: refactor how focused window and window group is tracked

Instead of recalculate every window's leader each time one of the
window's leader changed, aggregate all the changes and update all
the leaders in one go.

And now all the calculations are in one place too (wm/wm.c).

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
win: don't fetch window leader properties if not enabled

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: parse and use universal rules

Design is described in #1284.

Added a trivial test of universal rules

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
options: warn if both new and old style rules are used

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
docs: document universal window rules

This includes updating the changelog.

Closes #1284

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
rules: implement unredir window rules option

To replace wintypes::redir_ignore as well as
unredir-if-possible-exclude.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
c2: add new pre-defined target "group_focused"

This matches when the focused window is in the same window group as the
window being matched.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
docs: rewrite the "FORMAT OF CONDITIONS" section

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
docs: use anchors and xref

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
docs: formatting fixes

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

absolutelynothelix commented on man/picom.1.adoc:
```diff
@@ -275,47 +275,166 @@ May also be one of the predefined kernels: `3x3box` (default), `5x5box`, `7x7box
+	Whether to fade the matching window in/out when opening/closing it.
+
+  opacity:::
+	Opacity of the matching window. (0.0 - 1.0). If not explicitly set by a rule, the opacity value from the window properties (e.g. pass:[_]NET_WM_WINDOW_OPACITY)  will be used.
```
```suggestion
	Opacity of the matching window. (0.0 - 1.0). If not explicitly set by a rule, the opacity value from the window properties (e.g. pass:[_]NET_WM_WINDOW_OPACITY) will be used.
```

absolutelynothelix commented on man/picom.1.adoc:
```diff
@@ -275,47 +275,166 @@ May also be one of the predefined kernels: `3x3box` (default), `5x5box`, `7x7box
+	Corner radius of the matching window in number of pixels. 0 means no corner rounding.
+
+  blur-background:::
+	Whether to the background of the matching window should be blurred.
```
```suggestion
	Whether the background of the matching window should be blurred.
```

absolutelynothelix commented on man/picom.1.adoc:
```diff
@@ -275,47 +275,166 @@ May also be one of the predefined kernels: `3x3box` (default), `5x5box`, `7x7box
+	Whether to prevent the matching window from being painted over by shadows.
+
+  unredir:::
+	Whether the matching window should cause the compositor to unredirect the screen, and whether it should trigger the screen to be redirected again if it is currently unredirected. This could be a boolean value, if _true_, the screen will be unredirected if the matching window meets certain conditions; if _false_, it will never cause the screen to be unredirected. If the screen is currently unredi ...
```
```suggestion
	Whether the matching window should cause the compositor to unredirect the screen, and whether it should trigger the screen to be redirected again if it is currently unredirected. This could be a boolean value, if _true_, the screen will be unredirected if the matching window meets certain conditions; if _false_, it will never cause the screen to be unredirected. If the screen is currently unredirected, and there is no other window that will trigger unredirection, both of these choices will cause the screen to be redirected again. To control that behavior as well, you can set `unredir` to either _preferred_, such windows will not cause the screen to be redirected in this situation, and will behave like `true` otherwise; or _passive_, which not only won't cause redirection in this case, but also won't actively cause the screen to be unredirected. The last possible value for this option is _forced_, any of the windows having their `unredir` set to `forced` will cause the screen to be unredirected unconditionally. The value of the _pass:[_]NET_WM_BYPASS_COMPOSITOR_ property on the window will be considered if `unredir` is not explicitly set by any rule.
```

absolutelynothelix commented on man/picom.1.adoc:
```diff
@@ -275,47 +275,166 @@ May also be one of the predefined kernels: `3x3box` (default), `5x5box`, `7x7box
+Most of the rule options should 1:1 map to the new window rules. Here is a list of the non-trivial ones and how to achieve the same effect with window rules.
+
+*Inactive dimming and opacity*:: This includes options <<inactive-opacity>>, <<inactive-dim>>, <<active-opacity>>,
+<<inactive-opacity-override>>, <<mark-wmwin-focused>>, and <<mark-ovredir-focused>>. When using the window rules, the compositor no longer have an "active window" concept, as it is easy to achieve with window rules. You can use `match = "focused || group_focused"` to match windows that would have been considered active with the old options. Then you can set the opacity and dim level for matched w ...
```
```suggestion
<<inactive-opacity-override>>, <<mark-wmwin-focused>>, and <<mark-ovredir-focused>>. When using the window rules, the compositor no longer have an "active window" concept, as it is easy to achieve with window rules. You can use `match = "focused || group_focused"` to match windows that would have been considered active with the old options. Then you can set the opacity and dim level for matched windows accordingly. <<mark-wmwin-focused>> and <<mark-ovredir-focused>> can be achieved by adding `|| wmwin` and `|| override_redirect` to the match string, respectively. <<inactive-opacity-override>> can be achieved by setting `opacity-override = true`.
```

absolutelynothelix commented:
maybe it's a good idea to rewrite the sample config to use the `rules` option instead of all the old options

yshui committed:
docs: fix typos

Co-authored-by: Maxim Solovyov <msolovyov@protonmail.com>
Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
rules: implement full-shadow option for universal rules

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
wm/win: fix per-window fading configuration

When implementing animation, which also replaced the old fading code,
the per-window fading option was accidentally broken.

This commit brings it back.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
docs: update picom.sample.conf

Removed old options that are no long relevant (though not deprecated).

Converted all old rules to the new `rules` option.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>
